src/stats.ts

```diff
@@ -0,0 +1,187 @@
+import { Client } from './client';
+import { GenericObject, StatsEvent } from './types';
+import { isDfid } from './util/is';
+
+interface BaseStatsParams {
+  sessionId: string;
+  hashid?: string;
+}
+
+interface ClickStatsParams extends BaseStatsParams {
+  id: string;
+  datatype?: string;
+  query?: string;
+  customResultsId?: string | number;
+}
+
+interface BannerStatsParams extends BaseStatsParams {
+  bannerId: number;
+}
+
+export class StatsClient {
+  private client: Client;
+
+  public constructor(client: Client) {
+    this.client = client;
+  }
+
+  /**
+   * Allows changing the current client we
+   * are using
+   *
+   * @param  {Object}   client    The new Doofinder Client to use
+   *
+   */
+  public use(client: Client): void {
+    this.client = client;
+  }
+
+  /**
+   * Registers a new session with Doofinder, and returns
+   * the session ID generated for this session
+   *
+   * @param   {String}     sessionId   The sessionId
+   *
+   * @return  {String}     The newly generated Session ID
+   */
+  public async requestSession({ sessionId, hashid }: BaseStatsParams): Promise<GenericObject> {
+    if (!sessionId) {
+      throw new Error('Session ID must be defined');
+    }
+
+    if (hashid) {
+      this.client.hashid = hashid;
```

Comment:
Don't set the hashid in the client here…

Reply:
Why not? It basicly allows to set the hashid once and forget for the rest of the session, in case it is not passed around.

Reply:
If you want a default hashid you instantiate your client with it and just don't pass it as a parameter in the stats client. If you pass a hashid as a parameter, the client uses it, otherwise falls back to its own hashid. So this is not the right place and doing this kind of things in an _implicit_ way may be confusing (_why did my hashid changed in the client?_ you have to dig in the source code to know that the registerSession method changed it silently).